storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java

```diff
@@ -293,11 +293,11 @@ public ValidationResult assignOffsetsNonCompressed(LongRef offsetCounter,
 
         if (timestampType == TimestampType.LOG_APPEND_TIME) {
             maxTimestamp = now;
-            offsetOfMaxTimestamp = initialOffset;
-        }
-
-        if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
-            offsetOfMaxTimestamp = offsetCounter.value - 1;
+            if (toMagic >= RecordBatch.MAGIC_VALUE_V2) {
+                offsetOfMaxTimestamp = offsetCounter.value - 1;
+            } else {
+                offsetOfMaxTimestamp = initialOffset;
```

Comment:
We pass `offsetOfMaxTimestamp` to `ValidationResult.shallowOffsetOfMaxTimestampMs`. Using shallow offset used to be ok.  But when we added the `MAX_TIMESTAMP` support for ListOffset, it becomes important for offsetOfMaxTimestamp to be at the exact record level, instead of the shallow batch level. So, we need to rename `ValidationResult.shallowOffsetOfMaxTimestampMs` properly to make it clear. 

Also, there are other places like `LogSegment.append()` where we use the name `shallowOffsetOfMaxTimestamp`. We will need to rename it too to be clear.

Reply:
thanks for the comment! 
yes the naming is confusing, although it does represent the latest offset but it looks like it's only the offset of one batch, instead of the record. Let me address this.

Reply:
I have created a ticket [KAFKA-16380](https://issues.apache.org/jira/browse/KAFKA-16380) to track the changes of shallowOffsetOfMaxTimestamp for LogSegment